### ISSUE TITLE
Filter Rotation in Convolution 

### ISSUE THREAD

**Issue by MohdBaker** [fchollet/keras#2997]
Hi, 

Is there a way to rotate a filter in a convolution layer? Assuming I have only one filter, is there a way to configure the layer so that I get as an output the convolution of the input with 4 versions of the filter each rotated 90 degrees with respect to the prev. one? So my output is the output corresponding to 4 filters but these 4 filters are the rotations of one filter (they share the same rotated weights)

This can be done if I use 4 filters and find a way to force such constraint on the filter (if possible). 

Any help is appreciated

**Comment by fchollet** [comment]
Yes; you should reassign the layer's weights to be rotated versions of the previous weights. It isn't possible to do it as a step in the computation graph though.

But in between two steps, you can do:

```python
from keras import backend as K

w = K.get_value(layer.W)
w = rotate(w)
K.set_value(layer.W, w)
```

**Comment by MohdBaker** [comment]
I did not get that, Sorry!
let me give some more details:
I have a ConvNet, and I am using the built in layer Convolution2D with only 1 filter which is trainable but I want to have the result of this layer the convolution of the input with all rotated versions of the input. This means that instead of getting 1 convolution result fro the filter, I want 4

**Comment by MohdBaker** [comment]
I think I got it working, I updated the code for Convolution2D.. I am not sure it is the optimal way, but it is working!

Thanks

**Comment by MohdBaker** [comment]
I guess my fix did not work  since back-propagation will give error... I can evaluate the network but not train it (optimization). Can anyone help with this?

**Comment by tboquet** [comment]
I don't know if I understand your problem but you can maybe use a Lambda layer?
It's not tested but the idea is to rotate then concatenate the transformed input on the last axis as if you had more channels.
So you will have `number_of_rotations * number_of_channels` for each image.

```python
def rotate_input(x):
    x_1 = rotate(x)
    x_2 = rotate(x_1)
    ...
    return K.concatenate([x, x_1, ...], axis=2)

def rotate_input_output_shape(input_shape):
    shape = list(input_shape)
    shape[-1] *= number_of_rotations
    return tuple(shape)

model.add(Lambda(rotate_input, output_shape=rotate_input_output_shape))
```

I modified the [example](http://keras.io/layers/core/#lambda) from the doc.
Is it close to what you are looking for?

**Comment by MohdBaker** [comment]
Thanks, this is helpful, but what I am stuck at is how to rotate a tensor symbolically? 
So in you answer, you had x_1=rotate(x), I am not sure what is the function 'rotate'  that you are using? is it a tenor function?

**Comment by tboquet** [comment]
You should be able to flip your input in any of the dimension using 

```python
x[:, :, :, ::-1]
x[:, :, ::-1, :]
...
```

This way you can use combinations of flipping to rotate your arrays.

**Comment by MohdBaker** [comment]
Thank you!

**Comment by mbeach42** [comment]
@MohdBaker @tboquet  As far as I can see, this Lambda layer rotates the output on the filter, not the kernel itself. That's a different problem. Is there an elegant way to implement a conv2d which forces the 4 filters to be rotations of each other?

**Comment by YuanhaoGong** [comment]
indeed, this is a very powerful tool to force filters to be coupled. I am also interested the potential elegant implementation. Could someone share his/her implementation?

**Comment by YuanhaoGong** [comment]
@fchollet  @MohdBaker  @tboquet  @mbeach42  @bmabey Does anyone find a smart way to make the rotation version of the filter kernels? Getting the weight from one layer, rotating it, and then setting the weight is one option, but can NOT make these kernel coupled. For example, how to update them simultaneously? Someone might suggest shared layer. The problem here is "coupled kernels", not shared kernels. So, the shared layer is not a good option. 

In fact, the coupled kernels are very important in removing the redundancy. We could also use four independent kernels to do the task, which creases the parameter space. 

People augmented their data nowadays, but the coupled kernels are "augmented kernels". In that sense, "coupled rotation", "coupled flip", "coupled negative", etc. are more important than data augmentation.

It would be super cool if Keras can support kernel augmentation.

**Comment by redouanelg** [comment]
@mbeach42 @YuanhaoGong any success implementing this idea? I am interested too
